projects/packages/connection/tests/php/test_jetpack_xmlrpc_server.php

```diff
@@ -0,0 +1,438 @@
+		$this->assertEquals( sort( $user->allcaps ), sort( $response['allcaps'] ) );
+	}
+
+	/**
+	 * Test test_xmlrpc_get_user_by_id
+	 */
+	public function test_xmlrpc_remote_register_nonce_validation() {
+		$server  = new Jetpack_XMLRPC_Server();
+		$filters = array(
+			'return_invalid_nonce_status' => array(
+				'code'    => 400,
+				'message' => 'invalid_nonce',
+			),
+			'return_nonce_404_status'     => array(
+				'code'    => 400,
+				'message' => 'invalid_nonce',
+			),
+		);
+
+		foreach ( $filters as $filter => $expected ) {
+			add_filter( 'pre_http_request', array( $this, $filter ) );
+			$response = $server->remote_register(
+				array(
+					'nonce'      => '12345',
+					'local_user' => $this->xmlrpc_admin,
+				)
+			);
+			remove_filter( 'pre_http_request', array( $this, $filter ) );
+
+			$this->assertInstanceOf( 'IXR_Error', $response );
+			$this->assertEquals( $expected['code'], $response->code );
+
+			// Not using assertContains or assertStringContainsString to work with all phpunit versions.
+			$has_message = strpos( $response->message, sprintf( '[%s]', $expected['message'] ) );
+			$this->assertTrue( false !== $has_message );
+		}
+	}
+
+	/**
+	 * Test test_xmlrpc_get_user_by_id
+	 */
+	public function test_successful_remote_register_return() {
+		$server = new Jetpack_XMLRPC_Server();
+
+		// Set these so that we don't try to register unnecessarily.
+		Jetpack_Options::update_option( 'blog_token', 1 );
+		Jetpack_Options::update_option( 'id', 1001 );
+
+		add_filter( 'pre_http_request', array( $this, 'return_ok_status' ) );
+		$response = $server->remote_register(
+			array(
+				'nonce'      => '12345',
+				'local_user' => $this->xmlrpc_admin,
+			)
+		);
+		remove_filter( 'pre_http_request', array( $this, 'return_ok_status' ) );
+
+		$this->assertArrayHasKey( 'client_id', $response );
+		$this->assertEquals( 1001, $response['client_id'] );
+	}
+
+	/**
+	 * Test test_xmlrpc_get_user_by_id
+	 */
+	public function test_remote_connect_error_when_site_active() {
+		// Simulate the site being active.
+		Jetpack_Options::update_options(
+			array(
+				'blog_token' => 1,
+				'id'         => 1001,
+			)
+		);
+		( new Tokens() )->update_user_token( 1, sprintf( '%s.%d', 'token', 1 ), true );
+
+		$server = new Jetpack_XMLRPC_Server();
+
+		$response = $server->remote_connect(
+			array(
+				'nonce'      => '1234',
+				'local_user' => $this->xmlrpc_admin,
+			)
+		);
+
+		$this->assertInstanceOf( 'IXR_Error', $response );
+		$this->assertObjectHasAttribute( 'code', $response );
+		$this->assertObjectHasAttribute( 'message', $response );
+		$this->assertEquals( 400, $response->code );
+		$this->assertEquals(
+			'Jetpack: [token_fetch_failed] Failed to fetch user token from WordPress.com.',
+			$response->message
+		);
+
+		foreach ( array( 'blog_token', 'id', 'master_user', 'user_tokens' ) as $option_name ) {
+			Jetpack_Options::delete_option( $option_name );
+		}
```

Comment:
I note this cleanup won't happen if any of the tests fail. If the cleanup is actually needed, you should probably do something like this to make sure it happens.
```suggestion
		try {
			$server = new Jetpack_XMLRPC_Server();

			$response = $server->remote_connect(
				array(
					'nonce'      => '1234',
					'local_user' => $this->xmlrpc_admin,
				)
			);

			$this->assertInstanceOf( 'IXR_Error', $response );
			$this->assertObjectHasAttribute( 'code', $response );
			$this->assertObjectHasAttribute( 'message', $response );
			$this->assertEquals( 400, $response->code );
			$this->assertEquals(
				'Jetpack: [token_fetch_failed] Failed to fetch user token from WordPress.com.',
				$response->message
			);
		} finally {
			foreach ( array( 'blog_token', 'id', 'master_user', 'user_tokens' ) as $option_name ) {
				Jetpack_Options::delete_option( $option_name );
			}
		}
```